CMakeLists.txt

```diff
@@ -41,57 +41,92 @@ if(NOT DEFINED CMAKE_INSTALL_CMAKEDIR)
     set(CMAKE_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/nanopb")
 endif()
 
+# Determine Python module installation path
 if (NOT nanopb_PYTHON_INSTDIR_OVERRIDE)
     find_package(Python REQUIRED COMPONENTS Interpreter)
-    execute_process(
-        COMMAND ${Python_EXECUTABLE} -c
-            "import os.path, sys, sysconfig; print(os.path.relpath(sysconfig.get_path('purelib'), start=sys.prefix))"
-        OUTPUT_VARIABLE PYTHON_INSTDIR
-        OUTPUT_STRIP_TRAILING_WHITESPACE
-    )
+    file(TO_CMAKE_PATH "${Python_SITELIB}" PYTHON_INSTDIR)
 else()
     set(PYTHON_INSTDIR ${nanopb_PYTHON_INSTDIR_OVERRIDE})
 endif()
 message(STATUS "Python install dir: ${PYTHON_INSTDIR}")
 
+# Package nanopb generator as Python module 'nanopb'
 if(nanopb_BUILD_GENERATOR)
-    set(generator_protos nanopb)
-
-    foreach(generator_proto IN LISTS generator_protos)
-        string(REGEX REPLACE "([^;]+)" "${PROJECT_SOURCE_DIR}/generator/proto/\\1.proto" generator_proto_file "${generator_proto}")
-        string(REGEX REPLACE "([^;]+)" "\\1_pb2.py" generator_proto_py_file "${generator_proto}")
-        add_custom_command(
-            OUTPUT ${generator_proto_py_file}
-            COMMAND ${nanopb_PROTOC_PATH} --python_out=${PROJECT_BINARY_DIR} -I${PROJECT_SOURCE_DIR}/generator/proto ${generator_proto_file}
-            DEPENDS ${generator_proto_file}
-        )
-        add_custom_target("generate_${generator_proto_py_file}" ALL DEPENDS ${generator_proto_py_file})
-        install(
-            FILES ${PROJECT_BINARY_DIR}/${generator_proto_py_file}
-                  ${generator_proto_file}
-            DESTINATION ${PYTHON_INSTDIR}/proto/
-        )
-    endforeach()
+    # Copy Python code files related to the generator
+    add_custom_target(nanopb_generator ALL
+        COMMAND ${CMAKE_COMMAND} -E make_directory
+            ${PROJECT_BINARY_DIR}/nanopb/generator/proto
+
+        COMMAND ${CMAKE_COMMAND} -E copy_if_different
+            ${PROJECT_SOURCE_DIR}/generator/proto/_utils.py
+            ${PROJECT_SOURCE_DIR}/generator/proto/__init__.py
+            ${PROJECT_SOURCE_DIR}/generator/proto/nanopb.proto
+            ${PROJECT_BINARY_DIR}/nanopb/generator/proto
+
+        COMMAND ${CMAKE_COMMAND} -E copy_if_different
+            ${PROJECT_SOURCE_DIR}/generator/nanopb_generator.py
+            ${PROJECT_SOURCE_DIR}/generator/__init__.py
+            ${PROJECT_BINARY_DIR}/nanopb/generator
+
+        COMMAND ${CMAKE_COMMAND} -E copy_if_different
+            ${PROJECT_SOURCE_DIR}/generator/__init__.py
+            ${PROJECT_BINARY_DIR}/nanopb
```

Comment:
Why you copy this files to the binary directory? This has the disadvantage that you have to run cmake to update the files and could lead to problems if you remove a file in this list but not from the binary directory.

Reply:
It's so that running `nanopb_generator.py` from the build directory before installation works and uses the same directory structure. I'm not sure how much sense it makes to run it that way, but it seems like something someone will try to do and complain when it doesn't work.